### PR TITLE
feat: introduce JsInfo provider that encapsulates information provided by rules in rules_js and derivative rule sets

### PR DESCRIPTION
**Pull request by gregmagolan**
The changes are mostly under the hood and this should only be breaking to derivative rule sets and not to user code. PRs to update derivative rule sets will be queued up before this lands.

A new feature / pattern is introduced here to allow the propagation of linked npm deps to downstream npm_package targets so that they are direct deps when the package is linked with link_npm_package. This is done via the `data` attribute in `js_library`. Downstream rules that want this ability should use the same pattern. Helpers will be exported to make this easier.

Fixes https://github.com/aspect-build/rules_ts/issues/96

Re-does https://github.com/aspect-build/rules_js/pull/252 in a principled way that avoids the breakages in https://github.com/aspect-build/rules_js/pull/265 that that change caused

### CODE REVIEW

File: npm/private/npm_link_package.bzl [hunk 1]
```diff
@@ -289,22 +290,30 @@ deps of npm_link_package_store must be in the same package.""" % (ctx.label.pack
...
         # Always assume that packages provide typings, so we don't need to use an action to
-        # inspect the package.json#typings field or search for .d.ts files in the package.
-        declaration_info(
+        # inspect the package.json#typings field or search for .d.ts files in the package and
+        # so that any package can be passed as dep to rules that require DeclarationInfo such
+        # as ts_project.
+        DeclarationInfo(
             declarations = direct_files,
-            deps = ctx.attr.deps.keys(),
+            transitive_declarations = depset(transitive = [transitive_files] + transitive_declarations_depsets),
```
Comment (jbedard) [inline npm/private/npm_link_package.bzl]: Why are `transitive_files` in `transitive_declarations`? Is that just a naming thing?

File: npm/private/npm_link_package.bzl [hunk 2]
```diff
@@ -347,8 +347,9 @@ def _impl_direct(ctx):
         DefaultInfo(
             files = depset(output_files, transitive = [ctx.attr.src[DefaultInfo].files]),
         ),
-        declaration_info(
-            declarations = depset(output_files, transitive = [ctx.attr.src[DeclarationInfo].transitive_declarations]),
+        DeclarationInfo(
```
Comment (jbedard) [inline npm/private/npm_link_package.bzl]: 👍 for just directly using `DeclarationInfo` unless the factory method has actual logic in it, imo...

File: npm/private/npm_link_package.bzl [hunk 3]
```diff
@@ -292,25 +291,29 @@ deps of npm_link_package_store must be in the same package.""" % (ctx.label.pack
...
+    providers = [
+        DefaultInfo(files = transitive_files_depset),
         # Always assume that packages provide typings, so we don't need to use an action to
         # inspect the package.json#typings field or search for .d.ts files in the package and
         # so that any package can be passed as dep to rules that require DeclarationInfo such
         # as ts_project.
         DeclarationInfo(
-            declarations = direct_files,
-            transitive_declarations = depset(transitive = [transitive_files] + transitive_declarations_depsets),
+            declarations = files_depset,
```
Comment (jbedard) [inline npm/private/npm_link_package.bzl]: Should `declarations` not only be assigned declarations? Why the full `files_depset`?

Comment (jbedard) [reply npm/private/npm_link_package.bzl]: Same question in a few other places...

File: npm/private/npm_link_package.bzl [hunk 4]
```diff
@@ -336,26 +341,37 @@ def _impl_direct(ctx):
...
+            version = ctx.attr.src[NpmLinkedPackageStoreInfo].version,
+            store_info = ctx.attr.src[NpmLinkedPackageStoreInfo],
+            files = files_depset,
+            transitive_files = transitive_files_depset,
         ),
         DeclarationInfo(
-            declarations = depset(output_files, transitive = [ctx.attr.src[DeclarationInfo].declarations]),
-            transitive_declarations = ctx.attr.src[DeclarationInfo].transitive_declarations,
+            declarations = depset(files, transitive = [ctx.attr.src[DeclarationInfo].declarations]),
+            transitive_declarations = depset(files, transitive = [ctx.attr.src[DeclarationInfo].transitive_declarations]),
```
Comment (jbedard) [inline npm/private/npm_link_package.bzl]: `files` is both a declaraction + transitive declaration?

File: npm/private/npm_linked_package_direct_info.bzl [hunk 5]
```diff
@@ -0,0 +1,14 @@
+"NpmLinkedPackageDirectInfo provider"
+
+NpmLinkedPackageDirectInfo = provider(
+    doc = "Provides a direct linked npm package",
```
Comment (jbedard) [inline npm/private/npm_linked_package_direct_info.bzl]: what is a "direct linked npm package"?

File: npm/private/npm_link_package.bzl (hunk 1 above)
Comment (gregmagolan) [reply npm/private/npm_link_package.bzl]: It's a convention through rule sets. `foo` typically are direct foo things in a target and `transtive_foo` the direct foo things plus all the foo things from all deps & data.

Similar to `transitive_sources` in `PyInfo`L https://github.com/aspect-build/rules_py/blob/main/py/private/py_library.bzl#L98

File: npm/private/npm_link_package.bzl (hunk 2 above)
Comment (gregmagolan) [reply npm/private/npm_link_package.bzl]: I'm actually thinking of removing all our DeclarationInfo uses through the rule sets and carrying declarations in JsInfo. Working towards that here.

File: npm/private/npm_link_package.bzl (hunk 3 above)
Comment (gregmagolan) [reply npm/private/npm_link_package.bzl]: I think I broke GitHub. This diff view is out-of date and showing the wrong filename 🤷 . Guess I have too many commits?

File: npm/private/npm_linked_package_direct_info.bzl (hunk 5 above)
Comment (gregmagolan) [reply npm/private/npm_linked_package_direct_info.bzl]: Also out of date here. I'll ping you for review once I fix it. Will probably need a new PR

File: npm/private/npm_link_package.bzl (hunk 3 above)
Comment (gregmagolan) [reply npm/private/npm_link_package.bzl]: Oh, I force pushed the same commits and now it seems to be displaying correctly..

File: npm/private/npm_linked_package_direct_info.bzl (hunk 5 above)
Comment (gregmagolan) [reply npm/private/npm_linked_package_direct_info.bzl]: Force pushed the same commits and now it seems to be displaying correctly..

File: docs/js_binary.md [hunk 6]
```diff
@@ -60,6 +61,9 @@ This rules requires that Bazel was run with
...
 | <a id="js_binary-env"></a>env |  Environment variables of the action.<br><br>        Subject to <code>$(location)</code> and make variable expansion.   | <a href="https://bazel.build/docs/skylark/lib/dict.html">Dictionary: String -> String</a> | optional | {} |
 | <a id="js_binary-expected_exit_code"></a>expected_exit_code |  The expected exit code.<br><br>        Can be used to write tests that are expected to fail.   | Integer | optional | 0 |
+| <a id="js_binary-include_npm_deps"></a>include_npm_deps |  When True, npm_linked_deps from JsInfo and transitive_files in NpmLinkedPackageInfo and NpmPackageStoreInfo         providers in data targets are included in the runfiles of the target.   | Boolean | optional | True |
+| <a id="js_binary-include_transitive_declarations"></a>include_transitive_declarations |  When True, transitive_declarations from JsInfo providers in data targets are included in the runfiles of the target.<br><br>        Defaults to false since they are generally not needed at runtime and introducing them could slow down developer round trip         time due to having to generate typings on source file changes.   | Boolean | optional | False |
```
Comment (jbedard) [inline docs/js_binary.md:65]: > Defaults to false since they are

What is "they"? The providers? The data targets?

File: docs/js_library.md [hunk 7]
```diff
@@ -41,8 +41,8 @@ Can be used as a dep for rules that expect a DeclarationInfo such as ts_project.
 | Name  | Description | Type | Mandatory | Default |
 | :------------- | :------------- | :------------- | :------------- | :------------- |
 | <a id="js_library-name"></a>name |  A unique name for this target.   | <a href="https://bazel.build/docs/build-ref.html#name">Name</a> | required |  |
-| <a id="js_library-deps"></a>deps |  -   | <a href="https://bazel.build/docs/build-ref.html#labels">List of labels</a> | optional | [] |
-| <a id="js_library-srcs"></a>srcs |  -   | <a href="https://bazel.build/docs/build-ref.html#labels">List of labels</a> | optional | [] |
+| <a id="js_library-deps"></a>deps |  Dependencies of this library.<br><br>        This may include other js_library targets or targets that provider JsInfo as well as         linked npm dependencies.   | <a href="https://bazel.build/docs/build-ref.html#labels">List of labels</a> | optional | [] |
```
Comment (jbedard) [inline docs/js_library.md:44]: > linked npm dependencies

Should a provider be mentioned?

File: docs/js_library.md [hunk 8]
```diff
@@ -41,8 +41,8 @@ Can be used as a dep for rules that expect a DeclarationInfo such as ts_project.
...
+| <a id="js_library-srcs"></a>srcs |  The list of source files that are included in this library.<br><br>        This includes all your checked-in code and any generated source files.<br><br>        Other js_library targets and npm dependencies belong in <code>deps</code>.   | <a href="https://bazel.build/docs/build-ref.html#labels">List of labels</a> | optional | [] |
```
Comment (jbedard) [inline docs/js_library.md:45]: Does "checked-in" along with "any generated source files" exclude anything? Could we just say "source files"?

idk if mentioning "Other js_library targets" is really necessary either, especially since it was mentioned on the previous line 🤷

File: js/private/js_info.bzl [hunk 9]
```diff
@@ -0,0 +1,13 @@
+"JsInfo provider"
+
+JsInfo = provider(
+    doc = "Encapsulates information provided by rules in rules_js and derivative rule sets",
```
Comment (jbedard) [inline js/private/js_info.bzl:4]: is there an easy way to summarize what type of "information" or would that basically just be listing the params (source, declarations, npm links)?

File: js/private/js_library.bzl [hunk 10]
```diff
@@ -22,30 +22,45 @@ js_library(
...
 
 _DOC = """Copies all sources to the output tree and expose some files with DeclarationInfo.
 
 Can be used as a dep for rules that expect a DeclarationInfo such as ts_project."""
 
 _ATTRS = {
-    "srcs": attr.label_list(allow_files = True),
-    "deps": attr.label_list(allow_files = True),
+    "srcs": attr.label_list(
+        doc = """The list of source files that are included in this library.
```
Comment (jbedard) [inline js/private/js_library.bzl:35]: Drop "The list of"?

File: js/private/js_library.bzl [hunk 11]
```diff
@@ -68,48 +83,84 @@ target in {file_basename}'s package and add that target to the deps of {this_tar
...
+    declarations = []
+
     for src in output_srcs:
         if src.is_directory:
             # assume a directory contains typings since we can't know that it doesn't
-            typings.append(src)
+            declarations.append(src)
+            sources.append(src)
         elif (
             src.path.endswith(".d.ts") or
```
Comment (jbedard) [inline js/private/js_library.bzl:95]: I guess it's not really part of this PR but I think this also needs to include ".d.mts" and ".d.cjt" (https://github.com/aspect-build/rules_ts/commit/6a2841c4b6dd0760cc58d41e8762f5ab3833fb49)

File: js/private/js_library.bzl [hunk 12]
```diff
@@ -68,48 +83,84 @@ target in {file_basename}'s package and add that target to the deps of {this_tar
...
+    npm_linked_deps_depset = depset(transitive = npm_linked_deps_depsets)
 
     return [
+        JsInfo(
+            sources = sources_depset,
+            transitive_sources = transitive_sources_depset,
+            declarations = declarations_depset,
+            transitive_declarations = transitive_declarations_depset,
+            npm_linked_deps = npm_linked_deps_depset,
+            # npm_link_package_deps = ,
```
Comment (jbedard) [inline js/private/js_library.bzl:148]: ?

File: js/private/js_library.bzl [hunk 13]
```diff
@@ -68,48 +83,84 @@ target in {file_basename}'s package and add that target to the deps of {this_tar
...
         ),
-        OutputGroupInfo(types = typings),
+        OutputGroupInfo(types = declarations_depset),
     ]
 
 js_library_lib = struct(
     attrs = _ATTRS,
     implementation = _js_library_impl,
-    provides = [DefaultInfo, DeclarationInfo],
+    provides = [DefaultInfo, DeclarationInfo, JsInfo, OutputGroupInfo],
```
Comment (jbedard) [inline js/private/js_library.bzl:163]: `OutputGroupInfo` was missing? Does that mean we have no tests for it and should create a task to add tests?

File: js/private/js_run_binary.bzl [hunk 14]
```diff
@@ -176,16 +238,31 @@ def js_run_binary(
...
+    if hoist:
+        hoist_to_default_info_name = "{}_hoist_srcs_to_default_info".format(name)
+        _hoist_to_default_info(
+            name = hoist_to_default_info_name,
+            srcs = srcs,
+            include_transitive_sources = include_transitive_sources,
+            include_transitive_declarations = include_transitive_declarations,
+            include_npm_deps = include_npm_deps,
+            tags = kwargs.get("tags"),
+        )
```
Comment (jbedard) [inline js/private/js_run_binary.bzl:254]: 🤔 `testonly`?

File: js/private/js_run_binary.bzl [hunk 15]
```diff
@@ -176,16 +238,31 @@ def js_run_binary(
...
+        extra_srcs.append(":{}".format(hoist_to_default_info_name))
+
+    # Copy srcs to bin
     if copy_srcs_to_bin:
-        copy_to_bin_name = "%s_copy_srcs_to_bin" % name
+        copy_to_bin_name = "{}_copy_srcs_to_bin".format(name)
```
Comment (jbedard) [inline js/private/js_run_binary.bzl:259]: why `.format`?

File: js/private/js_library.bzl (hunk 13 above)
Comment (gregmagolan) [reply js/private/js_library.bzl:163]: I think that `OutputGroupInfo` and `DefaultInfo` are always implied and don't need to be specified

File: js/private/js_run_binary.bzl (hunk 15 above)
Comment (gregmagolan) [reply js/private/js_run_binary.bzl:259]: No good reason. I just find the `.format` express easier to read

File: docs/js_library.md (hunk 8 above)
Comment (gregmagolan) [reply docs/js_library.md:45]: > Thank you for doing many isolated commits 👍
> 
> Should we try to merge some of those sooner to avoid merge conflicts in this PR?

I've got some more work to do before any of it is ready. Will probably land as one big stack since it will be hard to reorder and squash this one.

File: js/private/js_info.bzl (hunk 9 above)
Comment (gregmagolan) [reply js/private/js_info.bzl:4]: Borrowed this text from PyInfo: https://docs.bazel.build/versions/main/skylark/lib/PyInfo.html#modules.PyInfo

Generic enough that it will never be out of date :)

File: js/private/js_library.bzl (hunk 12 above)
Comment (gregmagolan) [reply js/private/js_library.bzl:148]: WIP, I'm working up to that one which will fix https://github.com/aspect-build/rules_ts/issues/96

File: js/private/js_library.bzl (hunk 11 above)
Comment (gregmagolan) [reply js/private/js_library.bzl:95]: Oh right. There are more variants now.

File: npm/private/npm_link_package.bzl (hunk 1 above)
Comment (jbedard) [reply npm/private/npm_link_package.bzl]: But we're putting "files" into "declarations". Shouldn't declarations only be declaration files?

File: npm/private/npm_link_package.bzl (hunk 2 above)
Comment (jbedard) [reply npm/private/npm_link_package.bzl]: I'd want to think about that, fine grained vs all-in-one 🤔

Doesn't that mean if all you need is the `.d.ts` it might force other files to be generated? But having tons of tiny providers is a hassle and has its own problems of course.

File: npm/private/npm_link_package.bzl (hunk 1 above)
Comment (gregmagolan) [reply npm/private/npm_link_package.bzl]: This is out of date now that the code has been refactored further

File: npm/private/npm_link_package.bzl (hunk 2 above)
Comment (gregmagolan) [reply npm/private/npm_link_package.bzl]: This is out of date now that the code has been refactored further

File: npm/private/npm_link_package.bzl (hunk 3 above)
Comment (gregmagolan) [reply npm/private/npm_link_package.bzl]: This is out of date now that the code has been refactored further

File: npm/private/npm_link_package.bzl (hunk 4 above)
Comment (gregmagolan) [reply npm/private/npm_link_package.bzl]: This is out of date now that the code has been refactored further

File: npm/private/npm_linked_package_direct_info.bzl (hunk 5 above)
Comment (gregmagolan) [reply npm/private/npm_linked_package_direct_info.bzl]: This is out of date now that the code has been refactored further

Comment (gregmagolan) [reply npm/private/npm_linked_package_direct_info.bzl]: I dropped the "direct" nomenclature. There is now just a "link" or an "npm_package_store". A link is any package linked to node_modules `node_modules/<package>`. An "npm_package_store" is a package within the virtual store / symlinked node_modules tree `node_modules/.aspect_rules_js`

File: docs/js_binary.md (hunk 6 above)
Comment (gregmagolan) [reply docs/js_binary.md]: declarations. I updated the docstring to be more specific.

File: docs/js_library.md (hunk 7 above)
Comment (gregmagolan) [reply docs/js_library.md]: ✅

File: docs/js_library.md (hunk 8 above)
Comment (gregmagolan) [reply docs/js_library.md]: I borrowed the working from py_library srcs: https://docs.bazel.build/versions/main/be/python.html#py_library

Remove the "Other js_library targets" bit from the docstring.

File: js/private/js_library.bzl (hunk 10 above)
Comment (gregmagolan) [reply js/private/js_library.bzl]: ✅

File: js/private/js_library.bzl (hunk 11 above)
Comment (gregmagolan) [reply js/private/js_library.bzl]: I put this function in a central helper that downstream rules will use as well. We can add the additional variants in a follow-up.

Comment (gregmagolan) [reply js/private/js_library.bzl]: Actually. I just updated here to save time.

File: js/private/js_run_binary.bzl (hunk 14 above)
Comment (gregmagolan) [reply js/private/js_run_binary.bzl:219]: I created a new js_info_files rule that does this now

File: js/private/js_library.bzl (hunk 12 above)
Comment (gregmagolan) [reply js/private/js_library.bzl]: This is done now. See doctring in JsInfo for what it does.

**Comment by gregmagolan** [comment]
Went over this with @alexeagle and he suggested some nice improvements on some of this

File: js/libs.bzl [hunk 16]
```diff
@@ -0,0 +1,43 @@
+"""Starlark libraries for building derivate rules"""
```
Comment (alexeagle) [inline js/libs.bzl:1]: nit: derivative again

File: js/private/js_binary.bzl [hunk 17]
```diff
@@ -181,6 +148,24 @@ _ATTRS = {
...
+    ),
+    "include_declarations": attr.bool(
+        doc = """When True, 'declarations' and 'transitive_declarations' from 'JsInfo' providers in data targets are included in the runfiles of the target.
+
+        Defaults to false since declarations are generally not needed at runtime and introducing them could slow down developer round trip
+        time due to having to generate typings on source file changes.""",
+        default = False,
+    ),
+    "include_npm_linked_packages": attr.bool(
+        doc = """When True, files in 'npm_linked_packages' and 'transitive_npm_linked_packages' from 'JsInfo' providers in data targets are included in the runfiles of the target.
```
Comment (alexeagle) [inline js/private/js_binary.bzl:162]: how come you used single quotes instead of backticks?

File: js/private/js_library.bzl [hunk 18]
```diff
@@ -19,106 +18,124 @@ js_library(
...
+| Python sources and provides a `PyInfo`.
 """
 
-load("@aspect_bazel_lib//lib:copy_to_bin.bzl", "copy_files_to_bin_actions")
-load("@rules_nodejs//nodejs:providers.bzl", "DeclarationInfo", "declaration_info")
+load(":js_info.bzl", "JsInfo")
+load(":js_library_helpers.bzl", "gather_npm_linked_packages", "gather_npm_package_stores", "gather_runfiles", "gather_sources_and_declarations", "gather_transitive_declarations", "gather_transitive_sources")
+
+_DOC = """A library of JavaScript sources. Provides JsInfo, the primary provider used in rules_js
+and derivate rule sets.
```
Comment (alexeagle) [inline js/private/js_library.bzl:30]: search-replace for that misspelling?

File: js/private/js_library.bzl [hunk 19]
```diff
@@ -19,106 +18,124 @@ js_library(
...
+on declarations from 'deps' since these they don't need the JavaScript source files from deps to
+typecheck.
 
-_DOC = """Copies all sources to the output tree and expose some files with DeclarationInfo.
+Linked npm dependences are also handled separately from sources since not all rules require them and it
+is optimal for these rules to not depend on them in the build graph.
 
-Can be used as a dep for rules that expect a DeclarationInfo such as ts_project."""
+NB: 'js_library' copies all source files to the output tree before providing them in JsInfo. See
+https://github.com/aspect-build/rules_js/tree/67c561e08fa8baaf6fbfe00094fa0d757159e1ef#running-nodejs-programs
```
Comment (alexeagle) [inline js/private/js_library.bzl:41]: https://github.com/aspect-build/rules_js/tree/67c561e08fa8baaf6fbfe00094fa0d757159e1ef/docs#javascript might be a better link target

File: js/private/js_library_helpers.bzl [hunk 20]
```diff
@@ -0,0 +1,244 @@
...
+    # declarations
+    declarations.extend([
+        item
+        for target in targets
+        if JsInfo in target
+        for item in target[JsInfo].declarations
+    ])
+    declarations = [file for file in declarations if file]
+
+    return (sources, declarations)
```
Comment (alexeagle) [inline js/private/js_library_helpers.bzl:97]: do you have a reason for returning tuples rather than structs? the latter seems more common and idiomatic

File: js/private/js_binary.bzl (hunk 17 above)
Comment (gregmagolan) [reply js/private/js_binary.bzl:162]: I should fix vscode so you can just hit backtick and quote a selected word. Currently only works with single quote.

File: js/private/js_run_binary.bzl [hunk 21]
```diff
@@ -176,16 +186,33 @@ def js_run_binary(
...
+                                include_npm_linked_packages)
+    if make_js_filegroup_target:
+        js_filegroup_name = "{}_js_filegroup".format(name)
+        _js_filegroup(
+            name = js_filegroup_name,
+            srcs = srcs,
+            include_transitive_sources = include_transitive_sources,
+            include_declarations = include_declarations,
+            include_npm_linked_packages = include_npm_linked_packages,
+            tags = kwargs.get("tags"),
```
Comment (alexeagle) [inline js/private/js_run_binary.bzl:203]: should this use https://docs.aspect.build/aspect-build/bazel-lib/v1.5.0/docs/utils-docgen.html#propagate_well_known_tags rather than just "tags"?

File: js/private/js_library.bzl (hunk 18 above)
Comment (gregmagolan) [reply js/private/js_library.bzl:30]: ✅

File: js/libs.bzl (hunk 16 above)
Comment (gregmagolan) [reply js/libs.bzl:1]: ✅

File: npm/providers.bzl [hunk 22]
```diff
@@ -0,0 +1,18 @@
+"""Providers for building derivate rules"""
```
[the remark this answers is not in the document]
Comment (gregmagolan) [reply npm/providers.bzl:1]: ✅

File: npm/defs.bzl [hunk 23]
```diff
@@ -4,33 +4,11 @@
...
-)
 
-# export rules & macros as public API
 npm_package = _npm_package
 npm_link_package = _npm_link_package
-
-# export providers as public API
-NpmPackageInfo = _NpmPackageInfo
-NpmPackageStoreInfo = _NpmPackageStoreInfo
-NpmLinkedPackageInfo = _NpmLinkedPackageInfo
```
Comment (alexeagle) [inline npm/defs.bzl:33]: why no longer make the Providers public API?

File: js/private/js_library_helpers.bzl (hunk 20 above)
Comment (gregmagolan) [reply js/private/js_library_helpers.bzl:97]: good point, I'll switch to structs

File: js/private/js_run_binary.bzl (hunk 21 above)
Comment (gregmagolan) [reply js/private/js_run_binary.bzl:203]: in this case, I want to propagate all tags since I'm not looking for just exectution requirements. If a user tags the js_run_binary as manual then all targets it stamps out should get that too

File: npm/defs.bzl (hunk 23 above)
Comment (gregmagolan) [reply npm/defs.bzl:33]: moved to providers.bzl since only rules care about providers and you'll never see them in user BUILD files

**Review by alexeagle — APPROVED** [review]

File: js/private/js_library.bzl (hunk 19 above)
Comment (gregmagolan) [reply js/private/js_library.bzl]: ✅
